### PR TITLE
Support VMService.evaluate() for dart expressions

### PR DESCRIPTION
**Pull request by annagrin**
Support `VMService.evaluate` for dart expressions using expression evaluator.
This change allows evaluating arbitrary dart expressions in a current library, which allows clients like widget inspector to evaluate code needed to support layout inspector.
    
- Implement `ChromeProxyService.evaluate` API using expression compiler, with a fallback to the previous version if expression evaluation is turned off.
- Add module format to `ExpressionCompiler.initialize` API and to `LoadStrategy`.
- Add expression evaluation tests using `evaluate` API.
- Support sound null safety mode in frontend_server tests and add frontend server expression evaluation tests with sound null safety.
- Add webdev e2e test for `evaluate` API.

Note: this requires an SDK change that supports incremental compilation mode in DDC:
https://github.com/dart-lang/sdk/issues/41480

Incoming CR for the sdk issue:
https://dart-review.googlesource.com/c/sdk/+/188400

Closes: https://github.com/dart-lang/webdev/issues/1139

### CODE REVIEW

File: dwds/lib/src/loaders/strategy.dart [hunk 1]
```diff
@@ -28,6 +28,12 @@ abstract class LoadStrategy {
   /// This ID is passed to the injected client so that it can react accordingly.
   String get id;
 
+  /// Module format for JS generated by the compiler.
+  ///
+  /// Module format is passed to the expression compilation worker wuring
```
Comment (grouma) [inline dwds/lib/src/loaders/strategy.dart:33]: typo

File: dwds/lib/src/services/chrome_proxy_service.dart [hunk 2]
```diff
@@ -168,13 +168,15 @@ class ChromeProxyService implements VmServiceInterface {
 
   Future<void> updateCompilerDependencies(String entrypoint) async {
     var metadataProvider = globalLoadStrategy.metadataProviderFor(entrypoint);
+    var moduleFormat = globalLoadStrategy.moduleFormat;
```
Comment (grouma) [inline dwds/lib/src/services/chrome_proxy_service.dart:171]: Should the module format be a part of the metadata?

File: dwds/lib/src/services/chrome_proxy_service.dart [hunk 3]
```diff
@@ -345,14 +347,72 @@ ${globalLoadStrategy.loadModuleSnippet}("dart_sdk").developer.invokeExtension(
...
+          'Unrecognized isolate id: $isolateId. Supported isolate: ${isolate?.id}');
+    }
+  }
+
+  Future<Response> _getEvaluationResult(
+      Future<RemoteObject> Function() evaluation, String expression) async {
+    try {
+      var result = await evaluation();
+      // Handle compilation errors, internal errors,
+      // and reference errors from JavaScript evaluation in chrome
```
Comment (grouma) [inline dwds/lib/src/services/chrome_proxy_service.dart:363]: [nit] add period here and below.

File: dwds/lib/src/services/expression_compiler.dart [hunk 4]
```diff
@@ -67,7 +67,7 @@ abstract class ExpressionCompiler {
   /// Initializes the compiler with the provided null safety mode.
   ///
   /// May be called multiple times and always before [updateDependencies].
-  Future<void> initialize({bool soundNullSafety});
+  Future<void> initialize({String moduleFormat, bool soundNullSafety});
```
Comment (grouma) [inline dwds/lib/src/services/expression_compiler.dart:70]: Update the doc comment here.

File: dwds/lib/src/services/expression_evaluator.dart [hunk 5]
```diff
@@ -46,6 +46,78 @@ class ExpressionEvaluator {
...
+
+    // send js expression to chrome to evaluate
+
+    var isError = compilationResult.isError;
+    var jsExpression = compilationResult.result;
+
+    if (isError) {
+      // Frontend currently gives a text message including library name
+      // and function name on compilation error. Strip this information
+      // since it is shows syntetic names only used for debugger during
```
Comment (grouma) [inline dwds/lib/src/services/expression_evaluator.dart:75]: Some typos in this comment.

File: dwds/lib/src/services/expression_evaluator.dart [hunk 6]
```diff
@@ -46,6 +46,78 @@ class ExpressionEvaluator {
...
+        error = error.substring(0, error.lastIndexOf(']'));
+      }
+
+      if (error.contains('InternalError: ')) {
+        error = error.replaceAll('InternalError: ', '');
+        return _createError(ErrorKind.internal, error);
+      }
+
+      error = error.replaceAll(
+          RegExp('org-dartlang-debug:synthetic_debug_expression:.*:.*Error: '),
```
Comment (grouma) [inline dwds/lib/src/services/expression_evaluator.dart:95]: Can we make this a constant instead of constructing it for every eval?

File: frontend_server_common/lib/src/frontend_server_client.dart [hunk 7]
```diff
@@ -332,15 +310,25 @@ abstract class ResidentCompiler {
...
-  Future<CompilerOutput> compileExpression(
-    String expression,
-    List<String> definitions,
-    List<String> typeDefinitions,
-    String libraryUri,
-    String klass,
-    bool isStatic,
-  );
-
+  /// Compiles [expression] in [libraryUri] at[line]:[column] to JavaScript
```
Comment (grouma) [inline frontend_server_common/lib/src/frontend_server_client.dart:313]: [nit] space after `at`

File: webdev/test/e2e_test.dart [hunk 8]
```diff
@@ -276,115 +276,223 @@ void main() {
...
+
+              vmService = await vmServiceConnectUri(wsUri);
+              var vm = await vmService.getVM();
+              var isolate = await vmService.getIsolate(vm.isolates.first.id);
+              var library = isolate.libraries.first;
+
+              await vmService.streamListen('Debug');
+
+              var result = await vmService.evaluate(isolate.id, library.id,
+                  '(document?.body?.children[0] as SpanElement)?.text');
```
Comment (grouma) [inline webdev/test/e2e_test.dart:370]: Can we do `children.first` to test some extra logic?

File: dwds/lib/src/loaders/strategy.dart (hunk 1 above)
Comment (annagrin) [reply dwds/lib/src/loaders/strategy.dart:33]: thanks!

File: dwds/lib/src/services/chrome_proxy_service.dart (hunk 2 above)
Comment (annagrin) [reply dwds/lib/src/services/chrome_proxy_service.dart:171]: I thought about it a bit.  Currently for every scenario (flutter, google3, webdev) we have a preset constant format that can be just stored in the corresponding LoadStrategy, it is tightly coupled with other portions in there, like loadSnippet etc.

We could decouple load strategy from module format details in the future - store format in metadata and make the rest conditional on the format in any load strategy. This would be useful if we need to support multiple formats in one scenario.

File: dwds/lib/src/services/expression_evaluator.dart (hunk 6 above)
Comment (annagrin) [reply dwds/lib/src/services/expression_evaluator.dart:95]: sure, thanks

File: frontend_server_common/lib/src/frontend_server_client.dart (hunk 7 above)
Comment (annagrin) [reply frontend_server_common/lib/src/frontend_server_client.dart]: done, thanks

File: dwds/lib/src/services/expression_evaluator.dart (hunk 5 above)
Comment (annagrin) [reply dwds/lib/src/services/expression_evaluator.dart]: fixed, thanks

File: dwds/lib/src/services/expression_compiler.dart (hunk 4 above)
Comment (annagrin) [reply dwds/lib/src/services/expression_compiler.dart:70]: done, thanks

File: dwds/lib/src/services/chrome_proxy_service.dart (hunk 3 above)
Comment (annagrin) [reply dwds/lib/src/services/chrome_proxy_service.dart]: done, thanks

File: webdev/test/e2e_test.dart (hunk 8 above)
Comment (annagrin) [reply webdev/test/e2e_test.dart]: Sure, changed.

File: dwds/lib/src/services/chrome_proxy_service.dart (hunk 2 above)
Comment (grouma) [reply dwds/lib/src/services/chrome_proxy_service.dart:171]: Makes sense. Let's move forward with this for now.

**Review by grouma — APPROVED** [review]

**Comment by annagrin** [comment]
@grouma I disabled tests that depend on SDK changes for evaluate() support, so we can proceed with this PR. After my SDK changes propagate to dev channel,  I will re-enable the tests and update the min SDK.

**Review by grouma — APPROVED** [review]
